README.md
# PlotWaveform toolbox for Matlab audio data displayal

## Synposis
**PlotWaveform (PWF) is a Matlab tool to plot the waveform of a WAVE-file (or in newer Matlab versions even other formats using the ```audioread()``` function)using a block-by-block mean calculation algorithm. The number of blocks is defined by the size of the axis and therefore the available pixelwidth which enables the most precise rendering of the data while using less CPU cycles than with standard ```plot()``` generation. Additionally PWF supports a secondary display layer, in which audio data is plotted sample-exact just as the default ```plot()``` function would do. This layer is used only when a certain number of samples is not exceeded anymore. A third display layer supports the displayal of  sample-exact data in the views of ```stem()``` or ```stairs()``` functions. This view is set for the tightly zoomed audio portions.**

## Usage

__PlotWaveform__ is simply called with the filename of an audio file or with a vector audio samples in the first input argument. If executed with using a variable, the second input argument has to be the sampling rate. Either way numerous other function behavior modifiers are available (see list [below](#parameters)).

__WaveformPlayer__ is called the same way as PlotWaveform, only that it provides you with a basic audio player built right in. The integrated [Playrec](https://github.com/Janwillhaus/Playrec) audio playback and recording function is used to playback the displayed audio.

### Examples:
```Matlab
PlotWaveform('audiofile.wav');
PlotWaveform(audio_vec, fs);
PlotWaveform(__, 'ChannelView', 1);

WaveformPlayer('audiofile.wav');
WaveformPlayer(audio_vec, fs);
```

## Parameters

```Matlab
%   Possible function behavior settings:
%   ------------------------------------
%
%   'ChannelView':      1: creating a new plot for each channel in file
%                       0: all channels overlayed in one plot (default)
%
%   'ColorsetFace':     user defined vector of Colors to be used for the
%                       faces of the plot. For details see documentation
%
%   'ColorsetEdge':     user defined vector of Colors to be used for the
%                       edges of the plot. For details see documentation
%
%   'Interval':         user defined vector of time (start and end) in
%                       seconds in which the function should calculate
%
%   'PaperPosition':    user defined size for printing/exporting the 
%                       figure. 1x4 vector: [x_min y_min x_len y_len]
%
%   'PrintResolution':  user defined resolution of graph for printing/
%                       exporting. Possibilities: 150, 300, 600 (dpi)
%
%   'SilentPrint':      0: figure will stay open after print/export
%                       1: figure will be closed after print/export
%
%   'SampleViewStyle':  0: show sample-exact view as stems (default)
%                       1: show sample-exact view as stairs
%                       2: show sample-exact view as classic plot
%
%   'ShowXAxisAbove':   0: shows the x-axis below the plot (default)
%                       1: shows the x-axis above the plot
%
%   'PostZoomAction':   input function handle from superior function or
%                       script to be executed right after a zoom action
%                       was performed. Can be used as a trigger for
%                       additional modifications on the plot, made by the
%                       superior function or script.
%
%   'Axes':             user defined axes handle in which the waveform is
%                       desired to be plotted in. This is especially important
%                       if the function is used in multi-axes figures.
%
%   'Verbose':          integer to which to set the verbose logging. While
%                       future versions may offer more depth in verbose,
%                       right now only 0 (off) and 1 (on) are supported.
```

## Version History

Please consult the [CHANGELOG.md](CHANGELOG.md)

## License

**PlotWaveform** is available under X11 license

Copyright (c) 2011-2014 Jan Willhaus, Joerg Bitzer (Institute for Hearing Technology and Audiology at Jade University of Applied Sciences)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions: 

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# PLOTWAVEFORM Changelog

## v2.1 (03-Apr-2014)

* [PWF] Drastically improved the performance of PlotWaveform's data read-in. The processing is now "block-in-block", meaning only a slightly increased memory consumption as the drawback for amazing read-in times. As a benchmark, I tested a 6 hour wavefile, which was read in  under 40 seconds. 

* [WFP] Much better buffer handling of Playrec internally.

* [WFP] Added a settings panel in the Menubar (Audio -> Modify ...) to manipulate basic parameters of Playrec audio. For example, if the playback is dropping samples, increase the blocksize or GUI update interval.

* [WFP] Restored the broken functionality of the settings file for figure and audio settings in the player. If the function is called on its own (not integrated in another script's figure), the figure size and audio settings are kept til next opening.

[Commits](../../compare/2.0...2.1)


## v2.0 (26-Mar-2014)

* [WFP] Swapped Msound with [Playrec](https://github.com/Janwillhaus/playrec) and removed the binaries from this repository. With Playrec in Matlab path you'll now be able to use the playback functionality of the WaveformPlayer even on Macintosh! Make sure to get the most recent binaries of Playrec: https://github.com/Janwillhaus/playrec/releases/latest

* [WFP] [PWF] Moved the older changelogs from inside the function headers into CHANGELOG.md's bottom. WFP and PWF shall from now on use the same version number, making it more of a bundle's version number.

[Commits](../../compare/1.1...2.0)


## v1.1 (20-Mar-2014)

* A lot of minor fixes improving overall stability, including minor UI fixes

+ Changes are now documented in this very file (CHANGELOG.md), being outsourced from the functions head.

+ Support for the new `audioread()` function that was introduced to MATLAB. Thus Plotwaveform supports many more audio formats on newer MATLAB releases. Support for wavread is still present for older releases.

[Commits](../../compare/1.0...1.1)


## v1.0 (06-Sep-2013)

* Major release!

+ Finally added support for block-processing of wav read-in. PWF is now extremely fast in working with huge files; 1GB of wave can easily be loaded in about 15 seconds, while staying low in memory consumption.

+ Additional info: future versioning info will be placed inside the CHANGELOG.md file included in the repo. That way it is  easier to handle the multiple functions of the bundle. Cheers!

[Commits](../../compare/7b8ded3...1.0)


***
*Quoted below is the version history before moving it to CHANGELOG.md (in its original, but reversed order)*

**PlotWaveform**

```
%   Ver. 0.01   initial proof of concept script             10-Aug-2011     JB
%   Ver. 0.10   inital build of the function                27-Aug-2011     JW
%   Ver. 0.11   implementing mono & stereo view             31-Aug-2011     JW
%   Ver. 0.12   debugging second zoom layer                 01-Sep-2011     JW
%   Ver. 0.21   implementing two-axis view for stereo       07-Sep-2011     JW
%   Ver. 0.22   adding possibility to read multichannel     08-Sep-2011     JW
%   Ver. 0.23   improvements on multi-axis view             15-Sep-2011     JW
%   Ver. 0.24   debugging zoom layer switch                 22-Sep-2011     JW
%   Ver. 0.25   improvements on second zoom layer           28-Sep-2011     JW
%   Ver. 0.30   implementing customized zoom menu           12-Oct-2011     JW
%   Ver. 0.31   debugging customized zoom menu              23-Oct-2011     JW
%   Ver. 0.32   code cleaning, killing matlab warnings      01-Dec-2011     JW
%   Ver. 0.40   implementing print-functionhandle           30-Jan-2012     JW
%   Ver. 0.50   implementing wave-overview                  15-Feb-2012     JW
%   Ver. 0.51   fixed faulty zoom extract                   16-Feb-2012     JW
%   Ver. 0.60   supports PlotWaveformOverview function      27-Feb-2012     JW
%   Ver. 0.61   code cleaning, updating the help info       09-Mar-2012     JW
%   Ver. 0.70   code cleaning again. ready for public       01-Jun-2012     JW
%   Ver. 0.71   fixed some UI glitches due to use of gca    12-jan-2013     JW
%   Ver. 0.80   added 'axes' behavioral setting             12-Jan-2013     JW
%
%   Ver. 0.90   tons of improvements due to extended        25-Feb-2013     JW
%               usage in WaveformPlayer:
%               * fixed heavy glitches in terms of multi-
%                 axes and -figure use: The behavior of the
%                 'Axes' function property now works and
%                 allows the user to plot waveforms in any
%                 already given axes.
%               * reset to original values now works 
%                 correctly when ChannelView is off and 
%                 supports alpha blending in this state.
%               * other small bugfixes and improvements.
%
%   Ver. 0.91   Fixed the multi-axes behavior. PWF can     09-Mar-2013      JW
%               now be placed in axes even with channel
%               view activated. All channels will be 
%               plotted in place of the "parent" axes.
%
%   Ver. 1.0    Major release! PWF is now extremely fast   06-Sep-2013      JW
%               in working with huge files, due to a new
%               approach to block-processing the wavread
%               itself. 1GB of wave can easily be loaded
%               in about 15 seconds, while staying low in
%               memory consumption.
%               Additional info: future versioning info 
%               will be placed inside the CHANGELOG.md 
%               file included in the repo. That way it is 
%               easier to handle the multiple functions of 
%               the bundle. Cheers!
```



**WaveformPlayer**

```
%   Version History:
%   Ver. 0.01   initial creation of function                16-Jul-2012     JW
%   Ver. 0.10   first working build                         16-Jul-2012     JW
%   Ver. 0.20   basic functionality of playback             17-Jul-2012     JW

msoundPlayer.m
function stMsoundPlayer = msoundPlayer(varargin)
% Usage:
%   stMsoundPlayer = msoundPlayer({fs}, {blocklen})
%   stMsoundPlayer = msoundPlayer(WavFilename, {fs}, {blocklen})
%   stMsoundPlayer = msoundPlayer(inSamples, fs, {blocklen})
%
% Input:
%   WavFilename     Filename of Inputsignal
%   inSamples       Samples of Inputsignal
%   fs              Sampling Rate (default: 44100)
%   blocklen        Blocklen (default: 4096)
% Output:
%   stMsoundPlayer	MsoundPlayer-Object

% V 0.1    2011.05.03     S. Franz
if nargin == 0, help(mfilename); return; end;
if nargin > 3
    error('Only 3 Inputparameters (in, fs, blocklen) allowed!');
end
stSettings.in = [];
stSettings.fs = [];
stSettings.blocklen = [];

for iCount = 1 : length(varargin)
    var = cell2mat(varargin(iCount));
    if ischar(var)
        stSettings.in = var;
    elseif isnumeric(var)
        if length(var) == 1
            if isempty(stSettings.fs)
                stSettings.fs = var;
            elseif stSettings.fs < var
                stSettings.blocklen = stSettings.fs;
                stSettings.fs = var;
            else
                stSettings.blocklen = var;
            end
        else
            stSettings.in = var;
        end
    end
end
if isempty(stSettings.fs)
    stSettings.fs = 44100;
end
if isempty(stSettings.blocklen)
    stSettings.blocklen = stSettings.fs * (1024 * 4) / 44100;
end
stSettings.strPath = fileparts(mfilename('fullpath'));
stSettings.isPlaying = false;
stSettings.isPaused = false;
stSettings.firstStart = true;
stSettings.specLen = 2;
stSettings.specIdx = 0;
stSettings.maxSpecIdx = ceil(stSettings.fs / stSettings.blocklen) * stSettings.specLen;
stSettings.BlockProcess = 0;
stSettings.Samples = [];
stSettings.idx = [];
stSettings.currBlock = [];
stSettings.showSpecs = [1 1];
stSettings.showTimes = [1 1];
if exist([stSettings.strPath 'SpecSettings.mat'], 'file')
    showSpecs = [];
    showTimes = [];
    load([stSettings.strPath 'SpecSettings.mat']);
    stSettings.showSpecs = showSpecs;
    stSettings.showTimes = showTimes;
    clear showSpecs showTimes;
end

stMsoundPlayer.play = @play;
stMsoundPlayer.close = @close;
handles = struct();
handles.imagesc = struct();

init();

if ischar(stSettings.in)
    if exist(stSettings.in, 'file')
        [stSettings.Samples stSettings.fs] = wavread(stSettings.in);
        playButton();
    else
        error('File not found: %i', stSettings.in);
    end
elseif ~isempty(stSettings.in)
    stSettings.Samples = stSettings.in;
    playButton();
else
    set(handles.buttons.Play, 'enable', 'off');
    stSettings.BlockProcess = 1;
end

    function init()
        ImageSize = 16;
        handles.pictures.stop = zeros(ImageSize,ImageSize,3);
        handles.pictures.play = 0.9.*ones(ImageSize,ImageSize,3);
        handles.pictures.pause = zeros(ImageSize,ImageSize,3);
        handles.pictures.pause(:, 6 : 10, :) = .9;
        for count = 1:ImageSize/2
            handles.pictures.play(count,1:2*count,1) = zeros(1,2*count);
            handles.pictures.play(count,1:2*count,2) = zeros(1,2*count);
            handles.pictures.play(count,1:2*count,3) = zeros(1,2*count);
            handles.pictures.play(ImageSize+1-count,1:2*count,1) = zeros(1,2*count);
            handles.pictures.play(ImageSize+1-count,1:2*count,2) = zeros(1,2*count);
            handles.pictures.play(ImageSize+1-count,1:2*count,3) = zeros(1,2*count);
        end

        handles.figures.mainFigure.h = figure();
        set(handles.figures.mainFigure.h,  ...
            'Units', 'normalized', ...
            'resize','on',...
            'NumberTitle','off', ...
            'Position', [.1 .75 .8  .2] , ...
            'ToolBar', 'none', ...
            'KeyPressFcn', @onKeyPress, ...
            'MenuBar', 'none', ...
            'DeleteFcn', @closeWindow, ...
            'Name', 'msoundPlayer');

        handles.panel0 = uipanel('Title','Controls', 'Parent', handles.figures.mainFigure.h, 'position', [0 0 1/3 1], 'Units', 'normalized');

        handles.buttons.Play = uicontrol('Parent',handles.panel0,...
            'Units','Normalized',...
            'Position',[0/3+1/60 .7 1/3-2/60 .25],...
            'Style','PushButton',...
            'CData',handles.pictures.play, ...
            'enable', 'on', ...
            'Callback', @playButton);

        handles.buttons.Pause = uicontrol('Parent',handles.panel0,...
            'Units','Normalized',...
            'Position',[1/3+1/60 .7 1/3-2/60 .25],...
            'Style','PushButton',...
            'CData',handles.pictures.pause, ...
            'enable', 'off', ...
            'Callback', @pauseButton);

        handles.buttons.Stop = uicontrol('Parent',handles.panel0,...
            'Units','Normalized',...
            'Position',[2/3+1/60 .7 1/3-2/60 .25],...
            'Style','PushButton',...
            'CData',handles.pictures.stop, ...
            'enable', 'off', ...
            'Callback', @stopButton);

        handles.timeline = axes('Parent',handles.panel0,'Position',[1/60 .6 1-2/60 .05], 'ytick', [-.5 .5], 'yticklabel', {'' ''});
        plot(handles.timeline, [0 0] / stSettings.fs, [-.5 .5], 'r', 'LineWidth', 2);
        xlim(handles.timeline, [0 1]);
        set(handles.timeline, 'yticklabel', {});

        handles.panel1 = uipanel('Title','Channel 1', 'Parent', handles.figures.mainFigure.h, 'position', [1/3 0 1/3 1], 'Units', 'normalized');
        handles.spectogram1 = axes('parent', handles.panel1, 'OuterPosition', [0 0.1 1 .9]);
        hold(handles.spectogram1, 'on');
        handles.ChkSpectogram1 = uicontrol('Parent',handles.panel1,...
            'Units','Normalized',...
            'Position',[0 0 .5 .1],...
            'String', 'Spectogram', ...
            'Value', stSettings.showSpecs(1), ...
            'Style','Checkbox');

        handles.ChkTimesignal1 = uicontrol('Parent',handles.panel1,...
            'Units','Normalized',...
            'Position',[.5 0 .5 .1],...
            'String', 'Timesignal', ...
            'Value', stSettings.showTimes(1), ...
            'Style','Checkbox');

        handles.panel2 = uipanel('Title','Channel 2','Parent', handles.figures.mainFigure.h, 'position', [2/3 0 1/3 1], 'Units', 'normalized');
        handles.spectogram2 = axes('parent', handles.panel2, 'OuterPosition', [0 0.1 1 .9]);
        hold(handles.spectogram2, 'on');
        handles.ChkSpectogram2 = uicontrol('Parent',handles.panel2,...
            'Units','Normalized',...
            'Position',[0 0 .5 .1],...
            'String', 'Spectogram', ...
            'Value', stSettings.showSpecs(2), ...
            'Style','Checkbox');

        handles.ChkTimesignal2 = uicontrol('Parent',handles.panel2,...
            'Units','Normalized',...
            'Position',[.5 0 .5 .1],...
            'String', 'Timesignal', ...
            'Value', stSettings.showTimes(2), ...
            'Style','Checkbox');

        figure(handles.figures.mainFigure.h);
    end

    function isPlaying = play(in)
        if ishandle(handles.figures.mainFigure.h)
            set(handles.buttons.Stop, 'enable', 'on');
            stSettings.Samples = in;
            stSettings.SigLen = size(in, 1) / stSettings.fs;
            xlim(handles.timeline, [0 stSettings.SigLen]);
            stSettings.f = 0 : stSettings.fs/2;
            stSettings.t = 0 : 1 : stSettings.specLen;
            stSettings.idx = [0 size(in, 1)];
            if stSettings.firstStart == true
                msound('close');
                msound('openWrite', [], stSettings.fs, stSettings.blocklen, min(2, size(in, 2)), []);
                stSettings.firstStart = false;
                stSettings.isPlaying = true;
                stSettings.isPaused = false;
            end
            stSettings.currBlock = in(:, 1 : min(2, size(in, 2)));
            blockPlay();
            isPlaying = stSettings.isPlaying;
            drawnow;
        else
            isPlaying = false;
        end
    end

    function playButton(hObject, eventdata)
        if ~isempty(stSettings.Samples)
            set(handles.buttons.Play, 'enable', 'off');
            set(handles.buttons.Pause, 'enable', 'on');
            set(handles.buttons.Stop, 'enable', 'on');
            drawnow;
            if stSettings.isPlaying == false
                stSettings.SigLen = size(stSettings.Samples, 1) / stSettings.fs;
                xlim(handles.timeline, [0 stSettings.SigLen]);
                stSettings.isPlaying = true;
                stSettings.f = 0 : stSettings.fs/2;
                stSettings.t = 0 : 1 : stSettings.specLen;
                if isempty(stSettings.idx)
                    stSettings.idx = 1 : stSettings.blocklen;
                    if stSettings.firstStart == true
                        msound('close');
                        msound('openWrite', [], stSettings.fs, stSettings.blocklen, min(2, size(stSettings.Samples, 2)), []);
                        stSettings.firstStart = false;
                    end
                end
                while stSettings.isPlaying
                    if stSettings.isPaused == false
                        stSettings.currBlock = stSettings.Samples(stSettings.idx, 1 : min(2, size(stSettings.Samples, 2)));
                        blockPlay();
                        stSettings.idx = mod((stSettings.idx + stSettings.blocklen) - 1, size(stSettings.Samples, 1)) + 1;
                    end
                    drawnow;
                end
            else
                stSettings.isPaused = false;
            end
        end
    end

    function blockPlay()
        if stSettings.isPlaying && ishandle(handles.figures.mainFigure.h)
            set(get(handles.timeline, 'Children'), 'XData', [stSettings.idx(1) stSettings.idx(1)] / stSettings.fs);
            msound('putSamples', stSettings.currBlock);
            bFirst = false;
            if ishandle(handles.figures.mainFigure.h)
                if get(handles.ChkSpectogram1, 'Value') == 1 && size(stSettings.currBlock, 2) >= 1
                    mySpec = 20*log10(abs(spectrogram(stSettings.currBlock(:, 1), 'yaxis')) + eps);
                    if ~isfield(handles.imagesc, 'Spec1')
                        bFirst = true;
                        stSettings.spec1 = zeros(size(mySpec, 1), size(mySpec, 2) * stSettings.maxSpecIdx);
                        handles.imagesc.Spec1 = imagesc(stSettings.t, stSettings.f, stSettings.spec1, 'parent', handles.spectogram1, [-60 0]);
                        set(handles.spectogram1, 'YDir', 'normal');
                        if isfield(handles.imagesc, 'Time1')
                            delete(handles.imagesc.Time1);
                            handles.imagesc = rmfield(handles.imagesc, 'Time1');
                        end
                    end
                    stSettings.spec1(1:size(mySpec,1), (stSettings.specIdx * size(mySpec, 2) + (1 : size(mySpec, 2)))) = mySpec;
                    set(handles.imagesc.Spec1, 'CData', stSettings.spec1);
                elseif isfield(handles.imagesc, 'Spec1')
                    delete(handles.imagesc.Spec1);
                    handles.imagesc = rmfield(handles.imagesc, 'Spec1');
                end
                if get(handles.ChkSpectogram2, 'Value') == 1 && size(stSettings.currBlock, 2) >= 2
                    mySpec = 20*log10(abs(spectrogram(stSettings.currBlock(:, 2), 'yaxis')) + eps);
                    if ~isfield(handles.imagesc, 'Spec2')
                        bFirst = true;
                        stSettings.spec2 = zeros(size(mySpec, 1), size(mySpec, 2) * stSettings.maxSpecIdx);
                        handles.imagesc.Spec2 = imagesc(stSettings.t, stSettings.f, stSettings.spec2, 'parent', handles.spectogram2, [-60 0]);
                        set(handles.spectogram2, 'YDir', 'normal');
                        if isfield(handles.imagesc, 'Time2')
                            delete(handles.imagesc.Time2);
                            handles.imagesc = rmfield(handles.imagesc, 'Time2');
                        end
                    end
                    stSettings.spec2(1:size(mySpec,1), (stSettings.specIdx * size(mySpec, 2) + (1 : size(mySpec, 2)))) = mySpec;
                    set(handles.imagesc.Spec2, 'CData', stSettings.spec2);
                elseif isfield(handles.imagesc, 'Spec2')
                    delete(handles.imagesc.Spec2);
                    handles.imagesc = rmfield(handles.imagesc, 'Spec2');
                end
                if get(handles.ChkTimesignal1, 'Value') == 1 || get(handles.ChkTimesignal2, 'Value')
                    stepSize = 20;
                    currBlockTempBlock = stSettings.currBlock(1 : stepSize : end, :) * stSettings.fs / 4 + stSettings.fs / 4;
                    if ~isfield(stSettings, 'timeSignal') || isempty(stSettings.timeSignal)
                        stSettings.timeSignal = zeros(size(currBlockTempBlock, 1) * stSettings.maxSpecIdx, size(stSettings.currBlock, 2)) + stSettings.fs / 4;
                        if ~isfield(handles, 'timeSignal1') || ~ishandle(handles.timeSignal1)
                            handles.timeSignal1 = axes('parent', handles.panel1, 'Position',get(handles.spectogram1,'Position'), 'YAxisLocation','right', 'XTickLabel', [], 'Color','none', 'ylim', [-1 1]);
                        end
                    end
                    stSettings.timeSignal(stSettings.specIdx * size(currBlockTempBlock, 1) + (1 : size(currBlockTempBlock, 1)), :) = currBlockTempBlock;
                end
                if get(handles.ChkTimesignal1, 'Value') == 1 && size(stSettings.currBlock, 2) >= 1
                    if ~isfield(handles.imagesc, 'Time1')
                        bFirst = true;
                        handles.imagesc.Time1 = plot(handles.spectogram1, (0 : size(stSettings.timeSignal, 1) - 1) / stSettings.fs * stepSize, stSettings.timeSignal(:, 1), 'k');
                        if ~isfield(handles, 'timeSignal2') || ~ishandle(handles.timeSignal2)
                            handles.timeSignal2 = axes('parent', handles.panel2, 'Position',get(handles.spectogram2,'Position'), 'YAxisLocation','right', 'XTickLabel', [], 'Color','none', 'ylim', [-1 1]);
                        end
                    end
                    set(handles.imagesc.Time1, 'YData', stSettings.timeSignal(:, 1));
                elseif isfield(handles.imagesc, 'Time1')
                    delete(handles.imagesc.Time1);
                    handles.imagesc = rmfield(handles.imagesc, 'Time1');
                    stSettings.timeSignal(:, 1) = stSettings.timeSignal(:, 1) * 0 + stSettings.fs / 4;
                end
                if get(handles.ChkTimesignal2, 'Value') == 1 && size(stSettings.currBlock, 2) >= 2
                    bFirst = true;
                    if ~isfield(handles.imagesc, 'Time2')
                        handles.imagesc.Time2 = plot(handles.spectogram2, (0 : size(stSettings.timeSignal, 1) - 1) / stSettings.fs * stepSize, stSettings.timeSignal(:, 2), 'k');
                    end
                    set(handles.imagesc.Time2, 'YData', stSettings.timeSignal(:, 2));
                elseif isfield(handles.imagesc, 'Time2')
                    delete(handles.imagesc.Time2);
                    handles.imagesc = rmfield(handles.imagesc, 'Time2');
                    stSettings.timeSignal(:, 2) = stSettings.timeSignal(:, 2) * 0 + stSettings.fs / 4;
                end
                if bFirst
                    xlim(handles.spectogram1, [min(stSettings.t) max(stSettings.t)]);
                    ylim(handles.spectogram1, [min(stSettings.f) max(stSettings.f)]);
                    xlim(handles.spectogram2, [min(stSettings.t) max(stSettings.t)]);
                    ylim(handles.spectogram2, [min(stSettings.f) max(stSettings.f)]);
                end
                stSettings.specIdx = mod(stSettings.specIdx + 1, stSettings.maxSpecIdx);
            end
        else
            stopButton();
        end
    end

    function pauseButton(hObject, eventdata)
        set(handles.buttons.Play, 'enable', 'on');
        set(handles.buttons.Pause, 'enable', 'off');
        set(handles.buttons.Stop, 'enable', 'on');
        stSettings.isPaused = true;
    end

    function stopButton(hObject, eventdata)
        stSettings.isPlaying = false;
        stSettings.isPaused = false;
        stSettings.idx = [];
        if ishandle(handles.figures.mainFigure.h)
            if stSettings.BlockProcess == 0
                set(handles.buttons.Play, 'enable', 'on');
                set(handles.buttons.Pause, 'enable', 'off');
                set(handles.buttons.Stop, 'enable', 'off');
                if ~ischar(stSettings.in)
                    close();
                end
            else
                close();
            end
        end
    end

    function close()
        if ishandle(handles.figures.mainFigure.h)
            msound('close');
            delete(handles.figures.mainFigure.h);
        end
    end

    function onKeyPress(hObject, eventdata)
        if ishandle(handles.figures.mainFigure.h)
            if strcmp(eventdata.Key, 'x')
                close();
            elseif strcmp(eventdata.Key, 's')
                if strcmp(get(handles.buttons.Stop, 'enable'), 'on')
                    stopButton();
                end
            elseif strcmp(eventdata.Key, 'p')
                if strcmp(get(handles.buttons.Play, 'enable'), 'on')
                    playButton();
                elseif strcmp(get(handles.buttons.Pause, 'enable'), 'on')
                    pauseButton();
                end
            end
        end
    end

    function closeWindow(hObject, eventdata)
        if ishandle(handles.figures.mainFigure.h)
            showSpecs = [get(handles.ChkSpectogram1, 'Value') get(handles.ChkSpectogram2, 'Value')];
            showTimes = [get(handles.ChkTimesignal1, 'Value') get(handles.ChkTimesignal2, 'Value')];
            save([stSettings.strPath 'SpecSettings.mat'], 'showSpecs', 'showTimes');
            closereq();
        end
    end        
end

PlotWaveformTest.m
clear
close all
disp('----------------- Start Tester -----------------')

% szFileName = '/Users/janwillhaus/Documents/MATLAB/ichwillnur.wav';

% szFileName = 'SampleShort.wav';
% szFileName = 'SampleMid.wav';
szFileName = 'SampleLong.wav';
% szFileName = 'D:\Audio\Movie Audio\Katze_auf_dem_heissen_Dach_ita';
% szFileName = 'song4channels.wav';
% szFileName = 'songLeftLoud.wav';
% szFileName = 'SampleFlipChannel.wav';
% szFileName = 'SampleShort8channels.wav';
% szFileName = 'TomShort.wav';
% % % for input data vector testing
% xyz = wavread(szFileName);
% Out = PlotWaveform(xyz,44100);
    
% for multiple argument testing
% manualColorsetFace = [0.75 0.75 0.75; 0 0 0];
% manualColorsetEdge = [0.1 0.1 0.1];
% h = PlotWaveform(szFileName, 'ChannelView', 1, 'ColorsetFace', manualColorsetFace, 'ColorsetEdge', manualColorsetEdge);
% h = PlotWaveform(szFileName, 'ChannelView', 1, 'Interval', [1 3.5]);
% h = PlotWaveform(szFileName, 'ChannelView', 1, 'PrintResolution', 150);
% h = PlotWaveform(szFileName, 'ChannelView', 1, 'PaperPosition', [0 0 16 5], 'SilentPrint', 1,'PrintResolution', 150, 'Interval', [0 0.1]);

% testing internal print via function handle
tic;
[hFigure hAxes hPrint]= PlotWaveform(szFileName,...
    'ChannelView',1,  'SampleViewStyle',1, 'Verbose',1);
toc;
%[hFigure hAxes hPrint]= PlotWaveform(szFileName,'ChannelView',1, 'SampleViewStyle',1);
% hPrint(gcf,'-depsc',sprintf('%s-SampleViewStyleTest.eps',szFileName),'-painters','PaperPosition', [2 2 16 10],'SilentPrint',0) 

% h = PlotWaveform(szFileName, 'ChannelView', 1);

% for standard testing
% h = PlotWaveform(szFileName);

% szFileName = input('Enter path to local file\n')
% h = PlotWaveform(szFileName, 'ChannelView', 1);
disp('----------------- Stop Tester -----------------')


PlotWaveformOverviewTest.m
% Script to test the PlotWaveformOverview function
% Author: Jan Willhaus (c) IHA @ Jade Hochschule applied licence see EOF 
% Version History:
% Ver. 0.01 copy inherited from PlotWaveformTest    21-Feb-2012     JW

clear
close all
fprintf('\n\n\n.\n.\n.\n-------------- Start Tester --------------\n')
tic
% szFileName = 'SampleShort.wav';
%   szFileName = 'SampleMid.wav';
szFileName = 'SampleLong.wav';
%szFileName = 'song4channels.wav';
% szFileName = 'songLeftLoud.wav';
% szFileName = 'SampleFlipChannel.wav';
% szFileName = 'SampleShort8channels.wav';
% szFileName = 'TomShort.wav';

% % % for input data vector testing
% xyz = wavread(szFileName);
% Out = PlotWaveform(xyz,44100);


    
% for multiple argument testing
% manualColorsetFace = [0.75 0.75 0.75; 0 0 0];
% manualColorsetEdge = [0.1 0.1 0.1];
% h = PlotWaveform(szFileName, 'ChannelView', 1, 'ColorsetFace', manualColorsetFace, 'ColorsetEdge', manualColorsetEdge);
% h = PlotWaveform(szFileName, 'ChannelView', 1, 'Interval', [1 3.5]);
% h = PlotWaveform(szFileName, 'ChannelView', 1, 'PrintResolution', 150);
% h = PlotWaveform(szFileName, 'ChannelView', 1, 'PaperPosition', [0 0 16 5], 'SilentPrint', 1,'PrintResolution', 150, 'Interval', [0 0.1]);

% figure
% set(gcf, 'Position', [80 536 1760 420]);
% testing internal print via function handle
% [hFigure hAxes hOverview] = PlotWaveformOverview(szFileName);
PlotWaveformOverview(szFileName, 'ShowWaveOverview', 1, 'ShowPositionSliders', 1, 'Verbose',1);
%[hFigure hAxes hPrint]= PlotWaveform(szFileName,'ChannelView',1, 'SampleViewStyle',1);
% hPrint(gcf,'-depsc',sprintf('%s-SampleViewStyleTest.eps',szFileName),'-painters','PaperPosition', [2 2 16 10],'SilentPrint',0) 



T = toc;
fprintf('\nscript took %g seconds to finish.\n', T)
fprintf('\n-------------- Stop Tester --------------\n')

%--------------------Licence ---------------------------------------------
% Copyright (c) <2012> Jan Willhaus
% Institute for Hearing Technology and Audiology
% Jade University of Applied Sciences 
% Permission is hereby granted, free of charge, to any person obtaining 
% a copy of this software and associated documentation files 
% (the "Software"), to deal in the Software without restriction, including 
% without limitation the rights to use, copy, modify, merge, publish, 
% distribute, sublicense, and/or sell copies of the Software, and to
% permit persons to whom the Software is furnished to do so, subject
% to the following conditions:
% The above copyright notice and this permission notice shall be included 
% in all copies or substantial portions of the Software.
% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
% EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES 
% OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
% IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY 
% CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, 
% TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE 
% SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.